### PR TITLE
pkg: ccn_lite update

### PR DESCRIPTION
**Pull request by OlegHahm**
* updates CCN-lite to newest master
* enables RONR (reactive opportunistic name-based routing) in example application
* automatically use the first interface in the example application

**Update**
latest CCN-lite version provides:
* make the cache replacement strategy configurable
* utility function to print the content store
* allow loopback packets
* updated prototype for interest sending function
* fixes off-by one error for Interest retransmissions
* set the default cache size on RIOT to 5 (defined via `CCNL_CACHE_SIZE`)
* check the maximum number of supported interfaces (currently 1)
* limit the number of PIT entries (20 per default)
* some fixes for Interest timeout handling
* use xtimer instead of CCN-lite timer
* fix race condition in PIT timeout handling
* implement GNRC netapi send function

### CODE REVIEW

File: pkg/ccn-lite/ccn-lite-riot.h [hunk 1]
```diff
@@ -131,6 +131,23 @@ extern "C" {
...
 
+typedef struct {
+    struct ccnl_prefix_s *prefix;
+    unsigned char *buf;
+    size_t buflen;
+} ccnl_interest_t;
+
+
+/**
+ * Maximum number of elemtents that can be cached
```
Comment (aabadie) [inline pkg/ccn-lite/ccn-lite-riot.h]: elements (sorry, couldn't resist ;) k

Comment (OlegHahm) [inline pkg/ccn-lite/ccn-lite-riot.h]: That happens if you open a PR from a beergarden at IETF. ;)

Comment (OlegHahm) [inline pkg/ccn-lite/ccn-lite-riot.h]: fixed

File: pkg/ccn-lite/ccn-lite-riot.h [hunk 2]
```diff
@@ -164,17 +181,15 @@ int ccnl_open_netif(kernel_pid_t if_pid, gnrc_nettype_t netreg_type);
...
- * @param[in] name      The name that is requested
- * @param[in] chunknum  Number of the requested content chunk
+ * @param[in] prefix    The name that is requested
  * @param[out] buf      Buffer to write the content chunk to
  * @param[in] buf_len   Size of @p buf
  *
- * @return 0 on successfully sent Interest
- * @return -1 if Interested couldn't be sent
+ * @return pointer to the successfully sent Interest
+ * @return NULL if Interested couldn't be sent
```
Comment (aabadie) [inline pkg/ccn-lite/ccn-lite-riot.h]: Should it be `Interest` ? (it was already there)

Comment (OlegHahm) [inline pkg/ccn-lite/ccn-lite-riot.h]: thx, fixed

**Comment by kYc0o** [comment]
@all someone can test this one?

**Comment by emmanuelsearch** [comment]
I could test. I guess testing should start with: native, 3 nodes, chain topology. Right?

**Comment by OlegHahm** [comment]
I guess it's not necessary to test with 1000 nodes ;-), hence, yes, what you proposed should be sufficient.

**Comment by jia200x** [comment]
Hi.

I tried this example in M3 nodes and samr21-xpro, and they crash when I call `ccnl_open 3`.
In M3 the nodes stop responding.

In samr21 I got this:

2016-07-22 17:15:31,250 - INFO # > ifconfig
2016-07-22 17:15:31,254 - INFO # Iface  3   HWaddr: f5:d2  Channel: 26  Page: 0  NID: 0x23
2016-07-22 17:15:31,256 - INFO #            Long HWaddr: 5a:47:37:65:93:b1:f5:d2 
2016-07-22 17:15:31,263 - INFO #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2016-07-22 17:15:31,266 - INFO #            ACK_REQ  CSMA  
2016-07-22 17:15:31,270 - INFO #            Source address length: 2
2016-07-22 17:15:31,270 - INFO #            
ccnl_open 3
2016-07-22 17:15:35,116 - INFO # > ccnl_open 3
2016-07-22 17:15:35,973 - INFO # > Stack pointer corrupted, reset to top of stack
2016-07-22 17:15:35,975 - INFO # *** RIOT kernel panic:
2016-07-22 17:15:35,977 - INFO # HARD FAULT HANDLER
2016-07-22 17:15:35,977 - INFO # 
2016-07-22 17:15:35,985 - INFO # 	pid | name                 | state    Q | pri | stack ( used) | base       | current    
2016-07-22 17:15:35,992 - INFO # 	  - | isr_stack            | -        - |   - |   512 (  500) | 0x20000000 | 0x20000150
2016-07-22 17:15:36,000 - INFO # 	  1 | idle                 | running  Q |  15 |   256 (  132) | 0x20002598 | 0x20002614 
2016-07-22 17:15:36,011 - INFO # 	  2 | main                 | bl mutex _ |   7 |  1536 (  732) | 0x20002698 | 0x20002aec 
2016-07-22 17:15:36,017 - INFO # 	  3 | at86rf2xx            | bl rx    _ |   3 |  1024 (  332) | 0x2000557c | 0x2000587c 
2016-07-22 17:15:36,025 - INFO # 	  5 | ccnl                 | pending  Q |   6 |  1536 (  252) | 0x20000610 | 0x20000b14 
2016-07-22 17:15:36,033 - INFO # 	  5 | ccnl                 | pending  Q |   6 |  1536 (  252) | 0x20000610 | 0x20000b14 
2016-07-22 17:15:36,039 - INFO # 	    | SUM                  |            |     |  6400 ( 2200)
2016-07-22 17:15:36,040 - INFO # 
2016-07-22 17:15:36,041 - INFO # Heap usage:

I'm using the last commit of this PR (3df66db7827f0e6a8d9a401630e56fe98f96d95e). 
Am I missing something?

**Comment by OlegHahm** [comment]
Ah, sorry, forgot to update the README.

>   *  automatically use the first interface in the example application

Hence, no ccnl_open doesn't need to be called any more.

**Comment by jia200x** [comment]
Hi!

I could define the Interest and get it from other nodes. But for some reason when I execute ccnl_cont twice in the same node, I get a timeout.

E.g:

```
1469455269.487990;m3-101;> ccnl_int /riot/peter/schmerzl
1469455269.499614;m3-101;Content received: Hello World! Hello RIOT! 
m3-101;ccnl_int /riot/peter/schmerzl
1469455273.822026;m3-101;> ccnl_int /riot/peter/schmerzl
1469455293.822748;m3-101;timeout
m3-101;help                         
1469455313.824686;m3-101;timeout
1469455333.824666;m3-101;timeout
1469455333.826532;m3-101;Timeout! No content received in response to the Interest for /riot.
```
or
```
1469455586.938241;m3-101;> ccnl_int /riot/peter/schmerzl
1469455586.952945;m3-101;Content received: Hola mundo 
m3-100;ccnl_cont /riot/peter/schmerzl Hola mundo
1469455590.033918;m3-100;> ccnl_cont /riot/peter/schmerzl Hola mundo
m3-101;ccnl_int /riot/peter/schmerzl
```

Even if I generate again the content. The FIB route is the same in all cases:

```
1469455660.533891;m3-101;> ccnl_fib
1469455660.536377;m3-101;Prefix                         | Suite      | Interface | Peer
1469455660.539236;m3-101;-------------------------------|------------|-----------|------------------------------------
1469455660.540547;m3-101;/riot/peter/schmerzl           | ndn2013    |        03 | ff:ff:ff:ff:ff:ff/0x0801
```

Do you know what could be going on?
Cheers.

**Comment by OlegHahm** [comment]
I think this is somehow a problem of ccn-lite. The node will receive the content twice and notice that it has already cached it. Hence, it will discard the chunk. I think the valid behavior would be to look up in the local cache/content store first, but this needs to be addressed in a separate PR.

**Comment by OlegHahm** [comment]
Updated README and incorporate check for opening interface twice into the shell command correctly.

**Comment by kYc0o** [comment]
We need then to postpone this bug?

**Comment by OlegHahm** [comment]
Which one?

**Comment by kYc0o** [comment]
the bug addressed by this PR (I meant to postpone this PR)

**Comment by OlegHahm** [comment]
Hm, now, that I think about it: it's rather an enhancement than a bugfix. Fine to be postponed.

**Comment by kYc0o** [comment]
OK great!

**Comment by OlegHahm** [comment]
*bump*

**Comment by mfrey** [comment]
The removal of ``ccnl_open`` in the README is in "conflict with the example below the list of commands. I would propose a different example, but I already got an warning with the current example.

**Comment by OlegHahm** [comment]
> The removal of ccnl_open in the README is in "conflict" with the example below the list of commands. I would propose a different example

I still dislike too many examples in this repository, but no objections against another - maybe more advanced - example in the applications repository. For this example here, the question is to decide between displaying more configurability (with `ccnl_open`) or easier setup (with "auto-open").

> but I already got an warning with the current example.

Can you elaborate?

**Comment by mfrey** [comment]
> I still dislike too many examples in this repository, but no objections against another - maybe 

It is not about adding a new example, the old one simply doesn't make sense anymore if there is no ``ccnl_open`` command. If you can start using CCN-lite by just calling ``ccnl_cont`` maybe you can just remove the fourth entry of the example setup section. Or am I wrong? 

> Can you elaborate?

I don't think that this is the scope of this PR, but anyway:
```
main(): This is RIOT! (Version: 2016.10-devel-29-g39ecc-localhost)
Basic CCN-Lite example
> ifconfig    
ifconfig
Iface  3   HWaddr: 1a:d3:53:cc:10:af 
           
           Source address length: 6
           
> ccnl_open 3
ccnl_open 3
> ccnl_cont /riot/peter/schmerzl Hello World! Hello RIOT!
ccnl_cont /riot/peter/schmerzl Hello World! Hello RIOT!
 cache is full, cannot add new entry
```

**Comment by OlegHahm** [comment]
> It is not about adding a new example, the old one simply doesn't make sense anymore if there is no ccnl_open command.

Why do you think so? The current state of the example requires that the user specifies the interface to use for CCN. With this PR it just picks the first interface it can find.

> If you can start using CCN-lite by just calling ccnl_cont maybe you can just remove the fourth entry of the example setup section

You don't really *start* CCN-Lite with this command. You only add a content chunk to the node's content store. But yes, the fourth bullet point should be removed. I will update the PR.

About the problem you mention:
this should be fixed by this PR (by updating the CCN-lite version).
Alternatively, you can set `ccnl_relay.max_cache_entries = X` in your application with `X != 0`.

**Comment by mfrey** [comment]
> Why do you think so? The current state of the example requires that the user specifies the interface to use for CCN. With this PR it just picks the first interface it can find.

Well, the PR removes the ``ccnl_open`` command from the command list. The example uses it. Personally, I find this a bit misleading.

> About the problem you mention:

Will check.

**Comment by mfrey** [comment]
Within respect to the evaluation of the PR and @emmanuelsearch comment - chain refers to a simple grid and toying around with the example commands, or is there a more structured approach? Is there a way to test RONR?

**Comment by OlegHahm** [comment]
I think chain refers to a desvirt topology like
```
A   B   C
o---o---o
```

If you create content on node `A` and send an corresponding Interest from `C` you should get the content chunk only with RONR (or manually configured FIBs).  Following interests from `C` for the same prefix should be unicast (check with Wireshark).

**Comment by mfrey** [comment]
The file ``ndn.lua`` got changed so line 408 does not point anymore to
```
ethernetDissectorTable:add(0x8624, ndn)
```
but
```
pInfo.desegment_len = DESEGMENT_ONE_MORE_SEGMENT
```
I guess the line has to be updated in the README to [424](https://github.com/named-data/ndn-tools/blob/master/tools/dissect-wireshark/ndn.lua#L424)?

**Comment by mfrey** [comment]
> If you create content on node A and send an corresponding Interest from C you should get the content chunk only with RONR (or manually configured FIBs). Following interests from C for the same prefix should be unicast (check with Wireshark).

Did a quick test with a (A)--(B)--(C) topology, but need to verify it with wireshark. Will update the comment later with the wireshark output.

**Comment by OlegHahm** [comment]
Okay, updated to the latest CCN-lite version and updated the README.

**Comment by OlegHahm** [comment]
Updated the PR description with the CCN-lite changelog.

**Comment by mfrey** [comment]
I've did a quick test with 3x3 nodes (in a grid). The example seems to use RONR (because of the unicast communication). I've also added the pcap dump in case someone wants to check it. Any other tests I should perform? 
[pcap.zip](https://github.com/RIOT-OS/RIOT/files/424417/pcap.zip)

**Comment by OlegHahm** [comment]
Not from my side. Should I squash then?

**Comment by mfrey** [comment]
> Not from my side. Should I squash then?

:+1:

**Comment by OlegHahm** [comment]
done

**Comment by OlegHahm** [comment]
Anyone cares to ACK?

**Comment by emmanuelsearch** [comment]
Looks fine to me. ACK.

**Pull request merged**